enable other start values (#396)
openpharma/mmrm: vignettes/subsections/_intro-customizations.Rmd

to `mmrm`, or selected controls can be directly specified in the `mmrm` call.
We will see this below.

### Starting Values

The starting values will affect the optimization result. A better starting value
usually can make the optimization more efficient.
In `mmrm` we provide two starting value functions, one is `std_start` and the other is `emp_start`.

> `std_start` will try to use identity matrix as the covariance, however there are convergence problems

Reviewer: ```suggestion
`std_start` will try to use the identity matrix as the covariance, however there are convergence problems
```

> for `ar1` and `ar1h` if identity matrix is provided, thus for these two covariance structure

Reviewer: ```suggestion
for `ar1` and `ar1h` if the identity matrix is provided, thus for these two covariance structures
```

we use $\rho=0.5$ instead.

> `emp_start` will try to use empirical covariance as the starting value for unstructured covariance structure.

Reviewer: ```suggestion
`emp_start` will try to use the empirical covariance matrix of the residuals of the ordinary least squares model as the starting value for unstructured covariance structure.
```

If some timepoints are missing from data, identity matrix will be used for that submatrix.
The correlation between existing and non-existing timepoints are set to 0.

> As the starting values will affect the result, please be cautious on choosing the starting values.

Reviewer: I would say let's add here (the simplified form of) the motivating example to show that sometimes the default starting values don't work and you need to switch to emp_start

Author: added now

    sigmaa = 300,
    sigmab = 60,
    corab = 0.2,
    sigma = 10,
    times = c(0, 2, 6, 12, 24, 36, 52, 70, 88, 104)) {
  nt <- length(times)
  out <- data.frame(
    pts = rep(seq_len(n * 2), each = nt),
    trt = rep(c("Treatment", "Placebo"), rep(n * nt, 2)),
    time = rep(times, n * 2)
  )

  covab <- corab * sigmaa * sigmab # cov between a and b
  cov <- matrix(c(sigmaa^2, covab, covab, sigmab^2), ncol = 2) # Cov matrix for the slope and intercept
  si <- rbind(
    MASS::mvrnorm(n, mu = c(mua, mu + delta), Sigma = cov),
    MASS::mvrnorm(n, mu = c(mua, mu + delta), Sigma = cov)
  )
  idx <- rep(seq_len(n * 2), each = nt)
  out$fev1 <- si[idx, 1] + si[idx, 2] * times + rnorm(n * nt * 2, sd = sigma)
  out$trt <- factor(out$trt)
  out$time <- factor(out$time)
  out$pts <- factor(out$pts)
  return(out)
}
set.seed(123)
out <- gen_data()
```

In the generated data, the variance is not in the same scale across visits.

```{r show_variance}
vapply(split(out$fev1, out$time), sd, FUN.VALUE = 1)
```

Using `emp_start` as the starting value, `mmrm` will converge fast.
```{r mmrm_using_emp_start}
mmrm(fev1 ~ trt * time + us(time | pts), data = out, start = emp_start)
```

> However, if we use `std_start`, this will be convergence problems.

Reviewer: ```suggestion
However, if we use `std_start`, there will be convergence problems.
```

```{r mmrm_using_std_start, eval = FALSE}

> mmrm(fev1 ~ trt * time + us(time | pts), data = out, start = std_start)

Reviewer: can you add here the additional control options to increase number of iterations, as per the design doc? and add a comment before explaining that?
because it is also useful to see for the user

Author: added now

mmrm(fev1 ~ trt * time + us(time | pts), data = out, start = emp_start)
```

However, if we use `std_start`, there will be convergence problems.

> Here we also increase the maximum iteration number to 50000 to improve the convergence.

Reviewer: ```suggestion
We can also force a specific optimization algorithm and add control details, here e.g. choosing `nlminb` with increased maximum number of function evaluations and iterations.
```

> However this does not work and it gives errors.

Reviewer: I guess this will be slow but works?

Author: yes with nlminb it works

```{r mmrm_using_std_start, eval = FALSE}
mmrm(
  fev1 ~ trt * time + us(time | pts),
  data = out,
  start = std_start,

>   optimizer = "BFGS",
>   optimizer_control = list(maxit = 50000)

Reviewer: ```suggestion
   optimizer = "nlminb", 
   optimizer_control = list(eval.max = 1000, iter.max = 1000)
```
